lib/pci_caps.c

```diff
@@ -167,49 +167,36 @@ cap_write_pm(vfu_ctx_t *vfu_ctx, struct pci_cap *cap, char * buf,
 }
 
 static ssize_t
-handle_mxc_write(vfu_ctx_t *vfu_ctx, struct msixcap *msix,
-                 const struct mxc *const mxc)
+cap_write_msix(vfu_ctx_t *vfu_ctx, struct pci_cap *cap, char *buf,
+               size_t count, loff_t offset)
 {
-    assert(msix != NULL);
-    assert(mxc != NULL);
+    struct msixcap *msix = cap_data(vfu_ctx, cap);
+    struct msixcap new_msix = { { 0 } };
 
-    if (mxc->mxe != msix->mxc.mxe) {
-        vfu_log(vfu_ctx, LOG_DEBUG, "%s MSI-X",
-                mxc->mxe ? "enable" : "disable");
-        msix->mxc.mxe = mxc->mxe;
-    }
+    memcpy((char *)&new_msix + offset - cap->off, buf, count);
```

Comment:
don't we need to start with the old msix here, then copy over the new data?
what if we write to a byte that doesn't affect fm/mxe - won't that accidentally see them as zero?

Reply:
IIUC the only parts of the MSI-X capability structure that can be modified are the function mask and enable bits, the rest must ignored. In L190 and L196 we only store those bits.

Reply:
If the client does a single byte write at offset 0, that should indeed be ignored.
But then new_msix.mxc.fm and new_msix.mxe are zero, so you'll disable them incorrectly.

can add a test for this too.

Reply:
ah you're right